### ISSUE TITLE
Fetch leaks resources if body is not resolved. 

### ISSUE THREAD

**Issue by olaven** [denoland/deno#4735]
Hi 👋  
I believe I have found an issue where the `fetch`-api leaks resources if the 
body is not resolved, e.g. with `.text()`/`.json()`. 

Given the following test: 
```ts
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

Deno.test("fetching TODOs", async () => {

        const response = await fetch("https://jsonplaceholder.typicode.com/todos/");
        //await response.json(); //NOTE: adding this line removes the issue (or .text()/similar)
        assertEquals(200, response.status);
});
```

Produces the following output: 
```
running 1 tests
test fetching TODOs ... FAILED (94ms)

failures:

fetching TODOs
Error: Test case is leaking resources.
Before: {
  "0": "stdin",
  "1": "stdout",
  "2": "stderr"
}
After: {
  "0": "stdin",
  "1": "stdout",
  "2": "stderr",
  "3": "httpBody"
}
    at Object.assert ($deno$/util.ts:25:11)
    at Object.resourceSanitizer [as fn] ($deno$/testing.ts:75:5)
    at async TestApi.[Symbol.asyncIterator] ($deno$/testing.ts:259:11)
    at async Object.runTests ($deno$/testing.ts:341:20)

failures:

	fetching TODOs

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out (95ms)
```

The issue is only present from `v0.37.0`.

**Comment by syumai** [comment]
@olaven You should close response body.
Deno's fetch returns a `Response` with `Body` which implements `Deno.ReadCloser`.
`text()` and `json()` are calling [`Body.#bodyBuffer()`](https://github.com/denoland/deno/blob/0ea6eb83a906bff543be4c3301f23444986b022b/cli/js/web/fetch.ts#L47) which closes body implicitly.
Here's an example to resolve this issue.

```ts
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

Deno.test("fetching TODOs", async () => {
  const response = await fetch("https://jsonplaceholder.typicode.com/todos/");
  await response.body.close();
  assertEquals(200, response.status);
});
```

**Comment by olaven** [comment]
Thanks! I don't believe I have come across this before 🤔 
Does this apply to browser implementations as well?

**Comment by syumai** [comment]
It's not in browser's behavior. It's Deno's behavior (not in fetch standard).
Deno's fetch is more controllable than browser's fetch.
We can controll response body's resource explicitly.
Maybe Deno refers Go's HTTP implementation like this:

```go
resp, _ := http.Get("http://example.com/")
defer resp.Body.Close() // Body must be closed explicitly
```

However, I also think this could confuse JS land's programmers :sweat_smile:

**Comment by lucacasonato** [comment]
Well its a side effect of how `fetch` is implemented in Deno, but I think its a bug nonetheless because our fetch implementation should follow the spec, and this is not part of the spec.

**Comment by ry** [comment]
I've recently removed type declarations for `response.body.close()` as it's non-standard. I'm about to undertake a rewriting of fetch which will address these API issues. That said, even after this refactor, you will still need to consume the body in tests. Our tests check that resources are cleaned up. I suggest adding something like
```js
await response.arrayBuffer();
```
to fix the above test failure.

**Comment by alisabzevari** [comment]
As a followup question to this issue, wouldn't it be beneficial to close the resource after iterating over the body?

In other words, the following code fails as well:

``` ts
Deno.test({
  name: "test",
  fn: async () => {
    const response = await fetch("https://api.github.com/users/http4ts/repos");

    await readToEnd(response.body!);
  }
});

async function readToEnd(it: AsyncIterable<Uint8Array>) {
  let chunkCount = 0;
  for await (const chunk of it) {
    chunkCount += chunk.length;
  }
  console.log(`Finished reading ${chunkCount} bytes.`);
}
```

**Comment by 00ff0000red** [comment]
```js
const response = await fetch("https://tc39.github.io/ecma262");

for await (const value of response.body) {
	console.log(
		"Received %O bytes",
		value.length
	);
}

console.log("Response fully received");
```

Very similar to @alisabzevari's code.
I'm getting an OOM here, can't tell if it's because TC39's site is too large (too much memory allocated), or because of Deno's memory leakage.

**Comment by bartlomieju** [comment]
This is likely due to inefficient mechanism used in `fetch` API that JSON-stringifies payload, should be fixed by https://github.com/denoland/deno/pull/8826

**Comment by satyarohith** [comment]
> ```js
> const response = await fetch("https://tc39.github.io/ecma262");
> 
> for await (const value of response.body) {
> 	console.log(
> 		"Received %O bytes",
> 		value.length
> 	);
> }
> 
> console.log("Response fully received");
> ```
> 
> Very similar to @alisabzevari's code.
> I'm getting an OOM here, can't tell if it's because TC39's site is too large (too much memory allocated), or because of Deno's memory leakage, but it causes an OOM before finishing the loop, so I presume the former is the cause?

This issue no longer exists with `1.9.0` (Fixed by https://github.com/denoland/deno/pull/9036).

(cc @bartlomieju)

**Comment by CrimsonCodes0** [comment]
@satyarohith  That's good to hear, but has the primary issue here been fixed? I.e., are Response bodies correctly tracked and GCed?

**Comment by lucacasonato** [comment]
@CrimsonCodes0 No. This is a larger undertaking that requires some more core infrastructure to be built.

**Comment by evanwashere** [comment]
I am proposing to add unused fetch bodies limit alongside GC finalizers (based on [cloudflare workers design](https://developers.cloudflare.com/workers/platform/limits#simultaneous-open-connections))

This would allow runtime to cancel old **unread** response bodies if user set (or default) threshold is exceeded 

benefits:
- predictable memory usage
- less reliance on garbage collector
- fixed amount of leaking unread bodies (not yet garbage collected)

cons:
- runtime can cancel bodies which are going to be used later (rare/uncommon according to cloudflare docs)

example (6 leaking bodies allowed):
```js
const f1 = await fetch('https://example.com').then(res => res.status);
const f2 = await fetch('https://example.com').then(res => res.status);
const f3 = await fetch('https://example.com').then(x => x.body.getReader()); // or x.test/json/arrayBuffer() // body is used
const f4 = await fetch('https://example.com').then(res => res.status);
const f5 = await fetch('https://example.com').then(res => res.status);
const f6 = await fetch('https://example.com').then(res => res.status);
const f7 = await fetch('https://example.com').then(res => res.status);
const f8 = await fetch('https://example.com').then(res => res.status); // cancels f1 body
const f9 = await fetch('https://example.com').then(res => res.status); // cancels f2 body
const f10 = await fetch('https://example.com').then(res => res.status); // cancels f4 body
```

chart visualizing benefit
![chart](https://user-images.githubusercontent.com/70155278/122680747-c724c800-d1be-11eb-9f7d-c06c6f3ee677.png)
